internal/component/redis/redis.go

```diff
@@ -14,20 +14,62 @@ limitations under the License.
 package redis
 
 import (
-	"crypto/tls"
+	"context"
 	"fmt"
 	"strings"
 	"time"
 
-	"github.com/go-redis/redis/v8"
+	"golang.org/x/mod/semver"
 )
 
 const (
 	ClusterType = "cluster"
 	NodeType    = "node"
 )
 
-func ParseClientFromProperties(properties map[string]string, defaultSettings *Settings) (client redis.UniversalClient, settings *Settings, err error) {
+type RedisXMessage struct {
+	ID     string
+	Values map[string]interface{}
+}
+
+type RedisXStream struct {
+	Stream   string
+	Messages []RedisXMessage
+}
+
+type RedisXPendingExt struct {
+	ID         string
+	Consumer   string
+	Idle       time.Duration
+	RetryCount int64
+}
+
+type RedisPipeliner interface {
+	Exec(ctx context.Context) error
+	Do(ctx context.Context, args ...interface{})
+}
+
+var clientHasJSONSupport *bool
+
+type RedisClient interface {
+	Context() context.Context
+	DoResult(ctx context.Context, args ...interface{}) (interface{}, error)
+	DoErr(ctx context.Context, args ...interface{}) error
+	Close() error
+	PingResult(ctx context.Context) (string, error)
+	SetNX(ctx context.Context, key string, value interface{}, expiration time.Duration) (*bool, error)
+	EvalInt(ctx context.Context, script string, keys []string, args ...interface{}) (*int, error, error)
+	XAdd(ctx context.Context, stream string, maxLenApprox int64, values map[string]interface{}) (string, error)
+	XGroupCreateMkStream(ctx context.Context, stream string, group string, start string) error
+	XAck(ctx context.Context, stream string, group string, messageID string) error
+	XReadGroupResult(ctx context.Context, group string, consumer string, streams []string, count int64, block time.Duration) ([]RedisXStream, error)
+	XPendingExtResult(ctx context.Context, stream string, group string, start string, end string, count int64) ([]RedisXPendingExt, error)
```

Comment:
This method and its implementations in the two clients must be slow -- which is why PubSub cannot finish reading the remaining messages in time.

Would appreciate it if you can take a look at my conversion to the generic types.

@ItalyPaleAle

Reply:
Can you walk me through the problem tomorrow? I’m not sure I understand the issue.

Reply:
Conformance test now times out for PubSub. Only explanation can be my code that converts the v8 and v9 client specific types for messages and streams to my generic interface types.